avro_schema/frontend.lua

```diff
@@ -1243,11 +1243,66 @@ export_helper = function(node, already_built)
     end
 end
 
+local get_names_helper
+get_names_helper = function(res, pos, names, rec)
+    local fields = rec.fields
+    for i = 1, #fields do
+        local ftype = fields[i].type
+        insert(names, fields[i].name)
+        if type(ftype) == 'string' then
+            res[pos] = concat(names, '.')
+            pos = pos + 1
+        elseif ftype.type == 'record' and not ftype.nullable then
+            pos = get_names_helper(res, pos, names, ftype)
+        elseif not ftype.type then -- union
+            local path = concat(names, '.')
+            res[pos] = path .. '.$type$'
+            res[pos + 1] = path
+            pos = pos + 2
+        else
+            -- record*, scalar*, fixed, array, map
+            res[pos] = concat(names, '.')
+            pos = pos + 1
+        end
+        remove(names)
+    end
+    return pos
+end
+
+local get_types_helper
+get_types_helper = function(res, pos, rec)
+    local fields = rec.fields
+    for i = 1, #fields do
+        local ftype = fields[i].type
+        if type(ftype) == 'string' then
+            res[pos] = ftype
+            pos = pos + 1
+        elseif ftype.type == 'record' and not ftype.nullable then
+            pos = get_types_helper(res, pos, ftype)
+        elseif not ftype.type then -- union
+            res[pos] = "union_type"
+            res[pos + 1] = "union_value"
+            pos = pos + 2
+        else
+            -- record*, scalar*, fixed, array, map
+            local xtype = ftype.type
```

Comment:
The following code was here:

```
res[pos] = ftype.name or ftype.type
```

For what `ftype.name` was here? Don’t sure, but your changes seem good.

Reply:
I have decided that `ftype.name` was written there by mistake.
It leads to the problem shown (here)[https://github.com/tarantool/avro-schema/wiki/get_names-get_types-enhancement]. Fixed field is represented into `get_types` its name, not `field`, which is bad.

Reply:
`fixed_5` may be useful.
However, this `get_types` implementation is very limited. It does not support nested types and so on.
I think attributes like `size` and `nested` should be added in next API version and we do not need to invent a method to pass parameters.